src/runtime/MapObject.cpp

```diff
@@ -109,13 +119,47 @@ Value MapObject::getOrInsert(ExecutionState& state, const Value& key, const Valu
         }
     }
 
-    // If key is -0, let key be +0.
-    if (key.isNumber() && key.asNumber() == 0 && std::signbit(key.asNumber())) {
-        m_storage.pushBack(std::make_pair(Value(0), value));
-    } else {
-        m_storage.pushBack(std::make_pair(key, value));
+    m_storage.pushBack(std::make_pair(key, value));
+    return value;
+}
+
+Value MapObject::getOrInsertComputed(ExecutionState& state, Value& key, const Value& callback)
+{
+    if (!callback.isCallable()) {
+        ErrorObject::throwBuiltinError(state, ErrorCode::TypeError, "callback function is not callable");
+    }
+    key = canonicalizeKeyedCollectionKey(key);
+
+    for (size_t i = 0; i < m_storage.size(); i++) {
+        Value existingKey = m_storage[i].first;
+        if (existingKey.isEmpty()) {
+            continue;
+        }
+        if (existingKey.equalsToByTheSameValueZeroAlgorithm(state, key)) {
+            return m_storage[i].second;
+        }
+    }
+
+    Value argv[1] = { key };
+    Value value;
+    try {
+        value = Object::call(state, callback, Value(), 1, argv);
+    } catch (const Value& e) {
+        state.throwException(e);
+    }
```

Comment:
It seems like that this try-catch is unnecessary
It just invokes `callback` function and catches any exceptions, then simply re-throw it.
(catch statement doesn't do anything)
If an exception occurs during the execution of callback, then this exception will be thrown and delivered to the outer scope